src/data-sources/plugins/Constant/Settings.tsx

```diff
@@ -0,0 +1,130 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+
+import Chip from 'react-md/lib/Chips';
+import FontIcon from 'react-md/lib/FontIcons';
+import Paper from 'react-md/lib/Papers';
+import Divider from 'react-md/lib/Dividers';
+import TextField from 'react-md/lib/TextFields';
+import Button from 'react-md/lib/Buttons/Button';
+import SelectField from 'react-md/lib/SelectFields';
+
+import { BaseDatasourceSettings, IBaseSettingsProps, IBaseSettingsState } 
+                from '../../../components/common/BaseDatasourceSettings';
+
+export default class ConstantDatasourceSettings extends BaseDatasourceSettings<IBaseSettingsState> {
+
+  icon = 'chrome_reader_mode';
+  state = {
+    keyIndex: 0,
+    tokens: [],
+    newToken: '',
+    selectedToken: ''
+  };
+
+  constructor(props: IBaseSettingsProps) {
+    super(props);
+    this.removeToken = this.removeToken.bind(this);
+    this.addToken = this.addToken.bind(this);
+    this.onNewTokenChange = this.onNewTokenChange.bind(this);
+    this.selectedTokenChange = this.selectedTokenChange.bind(this);
+    this.state.keyIndex = 0;
+    this.state.tokens = this.props.settings.params['values'];
+    this.state.selectedToken = this.props.settings.params['selectedValue'];
+  }
+
+  removeToken(token: String) {
+    var tokens = this.state.tokens;
+    _.remove(tokens, function (x: String) {
+      return x === token;
+    });
+    this.setState({ tokens: tokens });
+    var selected = this.state.selectedToken;
+    
+    if (!_.find(tokens, function(x: any) {
+      return x === selected;
+    })) {
+      // the selected value was removed, thus, we set it to default
+      this.setState({ selectedToken: tokens[0] });
+    }
+  }
+
+  addToken() {
```

Comment:
This method should be:
addToken() {
  if (this.state.newToken) {
    **let** { tokens } = this.state;
    tokens = tokens || [];
    tokens.push(this.state.newToken);
    this.setState({ tokens, newToken: '' });
  }
}